### DISCUSSION TITLE
How to debug an INSERT when the call has an error?

### DISCUSSION THREAD

**Discussion by wsw70** [uptrace/bun#829] [Q&A]
I am a new user of Bun (and a false beginner in SQL for that matter), I am trying to use a rather complex `INSERT` that fails.

My database connection is initialized into a global variable used across my code:

```go
var db *bun.DB
var ctx context.Context

// setup DB connection
func init() {
	const dbFile = "data/db.sqlite"
	var err error

	sqldb, err := sql.Open(sqliteshim.ShimName, dbFile)
	if err != nil {
		log.Fatal().Msgf("cannot open database %s: %v", dbFile, err)
	} else {
		log.Info().Msgf("opened database %s", dbFile)
	}
	db = bun.NewDB(sqldb, sqlitedialect.New())

	// display queries on stdout in debug mode
	if os.Getenv("DEBUG") == "true" {
		db.AddQueryHook(bundebug.NewQueryHook(
			bundebug.WithVerbose(true),
			bundebug.FromEnv("BUNDEBUG"),
		))
	}

	// set context
	ctx = context.Background()
}
```

(I am still not sure what `ctx` is for (I could not find anything in the docs, but it works)

Another, related DB query (a helper database) does show the trace of the request for a `NewSelect()` - that query and the whole setup (similar to the one I showed above: `Open()`, `NewDB()`, ...) is within one function. The trace is as expected:

```
[bun]  10:38:50.944   SELECT              2.597377s  SELECT  FirstName, (...)
```

A `NewInsert()` query against `db` (the global variable) works but does not show the trace. It runs with an error 

```
cannot insert user TestOne: SQL logic error: near \"INSERT\": syntax error (1)
```

and I would be glad to see the whole `INSERT` to understand where the problem is.

**My question**: is the trace visible only after a successful request (a `nil` error returned)? If so: is there a way to debug the SQL command that is actually built?

EDIT: I do not see the trace even with a successful (error `nil`) request.

**Comment by bevzzz** [comment]
You can always:

```go
// Build the query first
q := db.NewInsert().Model(...).Set(...)

// Log the complete query to debug it
fmt.Printf(q.String())

// Execute later
if _, err := q.Exec(); err != nil {
  // handle error
}
```
